# Thông tin đồ án

## Thành viên

|   MSSV   |      Họ tên     |   GitHub  |
|:--------:|:---------------:|:---------:|
| 18120120 | Trần Chí Hào    | haotrch   |
| 18120147 | Trần Quốc Thắng | tqthangbl |

## Link thùng chứa GitHub

[https://github.com/tqthangbl/P4DS-FinalProject](https://github.com/tqthangbl/P4DS-FinalProject)

## Môi trường code

- Code trên môi trường Linux
- Dùng phiên bản các package như trong file "min_ds-env.yml"

In [1]:
import sys
sys.executable

'/home/family-u/.conda/envs/min_ds-env/bin/python'

## Import các thư viện cần thiết

In [2]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

# Thu thập dữ liệu

**Dữ liệu của bạn là về chủ đề gì và bạn lấy từ nguồn nào?**

Bộ dữ liệu mà nhóm sử dụng có tên là "Google Play Store Apps", được lấy từ [Kaggle](https://www.kaggle.com/lava18/google-play-store-apps), chứa thông tin về hơn 10000 ứng dụng trên Google Play Store. Thời điểm bộ dữ liệu được đăng lên Kaggle là vào ngày 05/09/2018, lần cập nhật cuối là vào 03/02/2019.

Có 2 file dữ liệu:
1. `googleplaystore.csv`: Chứa các thông tin chi tiết về các ứng dụng thu thập được từ Google Play.
2. `googleplaystore_user_reviews`: Chứa thông tin về 100 đánh giá liên quan nhất cho mỗi ứng dụng thu thập được từ Google Play.

**Người ta có cho phép bạn dùng dữ liệu như này không? Ví dụ, bạn có thể xem thử license của dữ liệu là gì**


License của dữ liệu là Creative Commons Attribution 3.0 Unported, theo như được nói ở [đây](https://creativecommons.org/licenses/by/3.0/), người ta cho phép:

1. Copy và chia sẻ dữ liệu này dưới bất kì hình thức nào
2. Sử dụng bộ dữ liệu này để xây dựng bất cứ thứ gì, cho bất cứ mục đích nào, kể cả thương mại

**Người ta thu thập dữ liệu như thế nào?**

Dữ liệu được tác giả crawl từ Google Play Store.


# Khám phá dữ liệu (đan xen với tiền xử lý)

## File `googleplaystore.csv`

Ở đây nhóm khám phá dữ liệu trong file `googleplaystore.csv` trước. Dữ liệu đọc từ file sẽ lưu vào dataframe `app_df`.

### Đọc dữ liệu từ file csv vào dataframe

In [3]:
app_df = pd.read_csv('Data/googleplaystore.csv')
app_df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [4]:
def get_app_df_shape():
    app_num_rows, app_num_cols = app_df.shape
    print(f'[+] num_rows = {app_num_rows}')
    print(f'[+] num_cols = {app_num_cols}')
    
get_app_df_shape()

[+] num_rows = 10841
[+] num_cols = 13


Dữ liệu có 10841 dòng và 13 cột.

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

In [5]:
app_df.sample(5)

App Category  Rating Reviews  Size  \
5848  Pinoy Kid Song Ako ay may Lobo   FAMILY     4.6      23   17M   
4541                              .R    TOOLS     4.5     259  203k   
8239          DB Ultra Saiyan Battle     GAME     4.5      33   44M   
6780                     ProGrade BT    TOOLS     NaN       0   46M   
2089      PJ Masks: Moonlight Heroes   FAMILY     4.4   86961   99M   

         Installs  Type Price Content Rating                     Genres  \
5848      10,000+  Free     0       Everyone                  Education   
4541      10,000+  Free     0       Everyone                      Tools   
8239       1,000+  Free     0   Everyone 10+                     Action   
6780         500+  Free     0       Everyone                      Tools   
2089  10,000,000+  Free     0       Everyone  Casual;Action & Adventure   

            Last Updated Current Ver   Android Ver  
5848  September 12, 2016         1.0    2.2 and up  
4541  September 16, 2014      1.1.06    1.5 and up  
8239       July 25, 2018       1.3.7    2.3 and up  
6780      April 27, 2018       1.0.2    4.3 and up  
2089        June 5, 2018       2.0.1  4.0.3 and up

Theo như mô tả của tác giả, mỗi dòng trong dữ liệu chứa thông tin của 1 ứng dụng trên Play Store. Bằng việc quan sát ngẫu nhiên các dòng trong dữ liệu nhiều lần, có vẻ không có dòng nào mang ý nghĩa khác.

### Mỗi cột có ý nghĩa gì?

Theo mô tả dữ liệu, các cột có ý nghĩa như sau:
* **App**: Tên ứng dụng
* **Category**: Hạng mục chính mà ứng dụng thuộc về
* **Rating**: Mức đánh giá tổng thể của ứng dụng (tại thời điểm thu thập dữ liệu)
* **Reviews**: Số lượng người đánh giá ứng dụng (tại thời điểm thu thập dữ liệu)
* **Size**: Kích thước ứng dụng (tại thời điểm thu thập dữ liệu)
* **Installs**: Số lượng người tải xuống / cài đặt ứng dụng (tại thời điểm thu thập dữ liệu)
* **Type**: Kiểu ứng dụng miễn phí (Free) hay trả phí (Paid)
* **Price**: Số tiền trả phí của ứng dụng (tại thời điểm thu thập dữ liệu)
* **Content Rating**: Nhóm độ tuổi mà ứng dụng nhắm đến
* **Genres**: Các thể loại mà ứng dụng thuộc về (có thể bao gồm cả hạng mục chính)
* **Last Updated**: Ngày cuối cùng mà ứng dụng được cập nhật trên Store (tại thời điểm thu thập dữ liệu)
* **Current Ver**: Phiên bản hiện tại của ứng dụng có mặt trên Store (tại thời điểm thu thập dữ liệu)
* **Android Ver**: Phiên bản Android yêu cầu tối thiểu (tại thời điểm thu thập dữ liệu)

### Dữ liệu có các dòng bị lặp không?

Xét dữ liệu có tồn tại các dòng bị lặp hay không:

In [6]:
app_df.duplicated().sum() > 0

True

#### Tiền xử lý loại bỏ các dòng trùng lặp

Do dữ liệu tồn tại các dòng bị lặp, và tuy là Play Store cho phép 2 ứng dụng có thể trùng tên, nhưng rất khó để 2 ứng dụng có thể trùng toàn bộ thông tin trong 12 cột còn lại. Do đó những dòng bị lặp này là của cùng 1 ứng dụng, vì vậy cần phải loại bỏ chúng.

In [7]:
app_df = app_df.drop_duplicates()
app_df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [8]:
get_app_df_shape()

[+] num_rows = 10358
[+] num_cols = 13


Số dòng của dữ liệu sau khi loại bỏ các dòng bị lặp, chỉ còn lại 10358 dòng.

#### Tiền xử lý loại bỏ các dòng trùng lặp giá trị cột `App`

Xem xét chi tiết hơn, theo mô tả dữ liệu, cột `App` có tất cả 9660 giá trị khác nhau. Do đó ta muốn kiểm tra rằng liệu những dòng có tên ứng dụng trùng nhau có phải là các ứng dụng khác nhau hay không.\
Xét một số dòng có giá trị cột `App` trùng nhau (bằng cách sắp xếp các dòng tăng dần theo cột `App` và giảm dần theo cột `Reviews`, xem 60 dòng đầu và 60 dòng cuối của bảng sau khi sắp xếp):

In [9]:
app_df[app_df.duplicated(subset='App', keep=False)]\
.sort_values(['App', 'Reviews'], ascending=[True, False]).head(60)

App             Category  \
3083                             365Scores - Live Scores               SPORTS   
5415                             365Scores - Live Scores               SPORTS   
1871                                         8 Ball Pool                 GAME   
1970                                         8 Ball Pool                 GAME   
1844                                         8 Ball Pool                 GAME   
1755                                         8 Ball Pool                 GAME   
1703                                         8 Ball Pool                 GAME   
1675                                         8 Ball Pool                 GAME   
3953                                         8 Ball Pool               SPORTS   
662                      95Live -SG#1 Live Streaming App               DATING   
559                      95Live -SG#1 Live Streaming App               DATING   
4991               A&E - Watch Full Episodes of TV Shows               FAMILY   
920                A&E - Watch Full Episodes of TV Shows        ENTERTAINMENT   
3799                       AC - Tips & News for Android™   NEWS_AND_MAGAZINES   
4928                       AC - Tips & News for Android™   NEWS_AND_MAGAZINES   
3791                           AP Mobile - Breaking News   NEWS_AND_MAGAZINES   
5448                           AP Mobile - Breaking News   NEWS_AND_MAGAZINES   
2800                                                ASOS             SHOPPING   
2771                                                ASOS             SHOPPING   
3628   AccuWeather: Daily Forecast & Live Weather Rep...              WEATHER   
5512   AccuWeather: Daily Forecast & Live Weather Rep...              WEATHER   
1143                        Acorns - Invest Spare Change              FINANCE   
1095                        Acorns - Invest Spare Change              FINANCE   
3524                                Adobe Acrobat Reader         PRODUCTIVITY   
3452                                Adobe Acrobat Reader         PRODUCTIVITY   
3895                                Adobe Acrobat Reader         PRODUCTIVITY   
2936   Adobe Photoshop Express:Photo Editor Collage M...          PHOTOGRAPHY   
2814   Adobe Photoshop Express:Photo Editor Collage M...          PHOTOGRAPHY   
1947                                             Agar.io                 GAME   
3880                                             Agar.io                 GAME   
3215                                              Airbnb     TRAVEL_AND_LOCAL   
3116                                              Airbnb     TRAVEL_AND_LOCAL   
2790        AliExpress - Smarter Shopping, Better Living             SHOPPING   
2660        AliExpress - Smarter Shopping, Better Living             SHOPPING   
2756        AliExpress - Smarter Shopping, Better Living             SHOPPING   
4099        AliExpress - Smarter Shopping, Better Living             SHOPPING   
2982                 All Football - Latest News & Videos               SPORTS   
5075                 All Football - Latest News & Videos               SPORTS   
10171                                All Social Networks               SOCIAL   
2549                                 All Social Networks               SOCIAL   
1933                                    Alto's Adventure                 GAME   
3936                                    Alto's Adventure                 GAME   
4083                                       Amazon Kindle  BOOKS_AND_REFERENCE   
143                                        Amazon Kindle  BOOKS_AND_REFERENCE   
2664                                     Amazon Shopping             SHOPPING   
2763                                     Amazon Shopping             SHOPPING   
3922                                     Amazon Shopping             SHOPPING   
2654                                  Amazon for Tablets             SHOPPING   
5353                                  Amazon for Tablets             SHOPPING   
315

In [10]:
app_df[app_df.duplicated(subset='App', keep=False)]\
.sort_values(['App', 'Reviews'], ascending=[True, False]).tail(60)

App         Category  \
4821                             Zello PTT Walkie Talkie           SOCIAL   
1255    Zomato - Restaurant Finder and Food Delivery App   FOOD_AND_DRINK   
1237    Zomato - Restaurant Finder and Food Delivery App   FOOD_AND_DRINK   
1982                                     Zombie Catchers             GAME   
1898                                     Zombie Catchers             GAME   
1924                                     Zombie Catchers             GAME   
1727                                     Zombie Catchers             GAME   
1693                                     Zombie Catchers             GAME   
4244                                     Zombie Catchers             GAME   
1948                                  Zombie Hunter King             GAME   
1874                                  Zombie Hunter King             GAME   
1728                                  Zombie Hunter King             GAME   
1669                                  Zombie Hunter King             GAME   
1914                                      Zombie Tsunami             GAME   
2008                                      Zombie Tsunami             GAME   
1730                                      Zombie Tsunami             GAME   
1685                                      Zombie Tsunami             GAME   
3940                                      Zombie Tsunami             GAME   
7077                      Zoosk Dating App: Meet Singles           DATING   
482                       Zoosk Dating App: Meet Singles           DATING   
3445                         ai.type Free Emoji Keyboard  PERSONALIZATION   
5863                         ai.type Free Emoji Keyboard  PERSONALIZATION   
9635    busuu: Learn Languages - Spanish, English & More           FAMILY   
835     busuu: Learn Languages - Spanish, English & More        EDUCATION   
808     busuu: Learn Languages - Spanish, English & More        EDUCATION   
2662   eBay: Buy & Sell this Summer - Discover Deals ...         SHOPPING   
4068   eBay: Buy & Sell this Summer - Discover Deals ...         SHOPPING   
743          edX - Online Courses by Harvard, MIT & more        EDUCATION   
789          edX - Online Courses by Harvard, MIT & more        EDUCATION   
389                            free video calls and chat    COMMUNICATION   
4224                           free video calls and chat    COMMUNICATION   
383                        imo free video calls and chat    COMMUNICATION   
347                        imo free video calls and chat    COMMUNICATION   
2663    letgo: Buy & Sell Used Stuff, Cars & Real Estate         SHOPPING   
4322    letgo: Buy & Sell Used Stuff, Cars & Real Estate         SHOPPING   
673    muzmatch: Muslim & Arab Singles, Marriage & Da...           DATING   
525    muzmatch: Muslim & Arab Singles, Marriage & Da...           DATING   
3492                                              myAT&T     PRODUCTIVITY   
4593                                              myAT&T     PRODUCTIVITY   
2331   mySugr: the blood sugar tracker made just for you          MEDICAL   
6194   mySugr: the blood sugar tracker made just for you          MEDICAL   
2595              ooVoo Video Calls, Messaging & Stories           SOCIAL   
2618              ooVoo Video Calls, Messaging & Stories           SOCIAL   
3334                             osmino Wi-Fi: free WiFi            TOOLS   
10473                            osmino Wi-Fi: free WiFi            TOOLS   
5664       realestate.com.au - Buy, Rent & Sell Property   HOUSE_AND_HOME   
1468       realestate.com.au - Buy, Rent & Sell Property   HOUSE_AND_HOME   
1891                                          slither.io             GAME   
1735                                          slither.io             GAME   
1757                                          slither.io             GAME   
1659                                          slither.io             GAME   
4227                                          sli

Theo quan sát sơ bộ, các dòng trùng nhau tại giá trị cột `App` đều có các thông tin giống nhau ở nhiều cột khác, nhưng khác nhau tối thiểu tại giá trị ở cột `Reviews`. Do đó với mỗi tên ứng dung, ta sẽ giữ lại dòng có giá trị ở cột `Reviews` cao nhất (có lẽ càng nhiều reviews sẽ cho thấy rằng dòng dữ liệu được thu thập ở thời gian mới nhất), và loại bỏ các dòng còn lại.

In [11]:
app_df = app_df.sort_values(['App', 'Reviews'], ascending=[True, False])\
                .drop_duplicates(subset='App').sort_index()
app_df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   
5                         Paper flowers instructions  ART_AND_DESIGN     4.4   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   
5     167  5.6M      50,000+  Free     0       Everyone   

                    Genres     Last Updated         Current Ver   Android Ver  
0             Art & Design  January 7, 2018               1.0.0  4.0.3 and up  
2             Art & Design   August 1, 2018               1.2.4  4.0.3 and up  
3             Art & Design     June 8, 2018  Varies with device    4.2 and up  
4  Art & Design;Creativity    June 20, 2018                 1.1    4.4 and up  
5             Art & Design   March 26, 2017                 1.0    2.3 and up

In [12]:
get_app_df_shape()

[+] num_rows = 9660
[+] num_cols = 13


Sau khi loại bỏ các dòng ở bước trên, bảng `app_df` có 9660 dòng và 13 cột.

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

Trước hết lấy ra `dtypes` của các cột, lưu vào series `col_dtypes`, và in ra để quan sát.

In [13]:
col_dtypes = app_df.dtypes
col_dtypes

App                object
Category           object
Rating            float64
Reviews            object
Size               object
Installs           object
Type               object
Price              object
Content Rating     object
Genres             object
Last Updated       object
Current Ver        object
Android Ver        object
dtype: object

Cột `Rating` thể hiện đánh giá của người dùng cho ứng dụng, là 1 con số nên kiểu `float64` là hợp lý. Các cột còn lại đang có kiểu `object`. Kiểu `object` trong Pandas thường là chuỗi, nhưng cũng có thể là một đối tượng bất kỳ. Do đó, trong các cột thuộc kiểu `object`, có thể chứa những giá trị thuộc các kiểu dữ liệu khác nhau. Ta phải kiểm tra điều này bằng cách: với mỗi cột, áp dụng lệnh `type` lên từng phần tử trong cột đó để xem kiểu dữ liệu thật sự của chúng. Hàm `open_object_dtype` sẽ nhận input là 1 series `s`, trả về 1 set là tập các kiểu dữ liệu có trong series đó. Khi xem kết quả hàm `open_object_dtype` của từng cột, ta cũng xem luôn cột đó mang các giá trị phân biệt nào để xem có điều gì bất thường hay không.

In [14]:
def open_object_dtype(s):
    return set(s.apply(type).unique())

for col_name in col_dtypes.index:
    if col_dtypes[col_name] == 'object':
        print(f'[+] Column {col_name} dtypes: ')
        print(open_object_dtype(app_df[col_name]))
        print(app_df[col_name].unique())
        print()

[+] Column App dtypes: 
{<class 'str'>}
['Photo Editor & Candy Camera & Grid & ScrapBook'
 'U Launcher Lite – FREE Live Cool Themes, Hide Apps'
 'Sketch - Draw & Paint' ... 'Parkinson Exercices FR'
 'The SCP Foundation DB fr nn5n'
 'iHoroscope - 2018 Daily Horoscope & Astrology']

[+] Column Category dtypes: 
{<class 'str'>}
['ART_AND_DESIGN' 'AUTO_AND_VEHICLES' 'BEAUTY' 'BOOKS_AND_REFERENCE'
 'BUSINESS' 'COMICS' 'COMMUNICATION' 'DATING' 'EDUCATION' 'ENTERTAINMENT'
 'EVENTS' 'FINANCE' 'FOOD_AND_DRINK' 'HEALTH_AND_FITNESS' 'HOUSE_AND_HOME'
 'LIBRARIES_AND_DEMO' 'LIFESTYLE' 'GAME' 'FAMILY' 'MEDICAL' 'SOCIAL'
 'SHOPPING' 'PHOTOGRAPHY' 'SPORTS' 'TRAVEL_AND_LOCAL' 'TOOLS'
 'PERSONALIZATION' 'PRODUCTIVITY' 'PARENTING' 'WEATHER' 'VIDEO_PLAYERS'
 'NEWS_AND_MAGAZINES' 'MAPS_AND_NAVIGATION' '1.9']

[+] Column Reviews dtypes: 
{<class 'str'>}
['159' '87510' '215644' ... '603' '1195' '398307']

[+] Column Size dtypes: 
{<class 'str'>}
['19M' '8.7M' '25M' '2.8M' '5.6M' '29M' '33M' '3.1M' '28M' '12M

#### Vấn đề cần tiền xử lý: Tồn tại dòng dữ liệu bị thiếu giá trị cột nào đó

Quan sát kết quả, khoan nói đến kiểu dữ liệu ở từng cột, ta thấy có một số điểm bất thường khi xem các giá trị phân biệt ở từng cột:
- Ở cột `Category`, có giá trị là `1.9`, có vẻ không phải là 1 giá trị phù hợp với ý nghĩa category
- Ở cột `Size`, có giá trị là `1,000+`, so với các giá trị còn lại trong cột này thường có format là 1 con số, theo sau đó là đơn vị (`k` hoặc `M`). Giá trị `1,000+` này có vẻ phù hợp với format của cột ngay sau đó là `Install` hơn.
- Ở cột `Installs`, có 1 giá trị là `Free`, không nói lên số lượt cài đặt của ứng dụng, giá trị này lại phù hợp với cột tiếp theo là `Type`, mang ý nghĩa ứng dụng đó thuộc loại miễn phí hay trả phí thì đúng hơn.

Từ đó ta đoán rằng có một dòng nào đó trong dữ liệu bị thiếu giá trị một số cột, do đó các giá trị của các cột còn lại bị dịch qua trái, nên giá trị ở một cột này lại nằm ở cột ngay trước đó. Đọc phần [Discussion](https://www.kaggle.com/lava18/google-play-store-apps/discussion/164101) trên Kaggle cũng có người gặp vấn đề này, họ nói rằng dòng 10472 bị thiếu giá trị. Nhưng ta vẫn cần tiền xử lý để tìm và sửa các dòng như vậy, cho chắc.

#### Tìm các dòng có giá trị một số cột bị thiếu

Ta sử dụng thư viện `csv` để đọc file `googleplaystore.csv`, dòng đầu của file sẽ là tên các cột. Với mỗi dòng tiếp theo, nếu số cột trong dòng đó khác với số lượng cột thì đưa chỉ số của dòng vào list `bad_index`. Sau khi chạy code, `bad_index` sẽ chứa các chỉ số dòng có các cột bị thiếu.

In [15]:
import csv

csv_file = open('Data/googleplaystore.csv')
csv_reader = csv.reader(csv_file, delimiter=',')
row_id = -1
col_count = 0
bad_index = []

for row in csv_reader:
    if row_id == -1:
        col_count = len(row)
        print(f"CSV file has {col_count} columns")
    else:
        if len(row) != col_count:
            print(f"Row id = {row_id} has {len(row)} columns")
            bad_index.append(row_id)
    row_id += 1

CSV file has 13 columns
Row id = 10472 has 12 columns


#### Thay thế các giá trị cột bị thiếu thành `NaN`

Xét dòng bị thiếu cột:

In [16]:
app_df.loc[bad_index]

App Category  Rating Reviews  \
10472  Life Made WI-Fi Touchscreen Photo Frame      1.9    19.0    3.0M   

         Size Installs Type     Price Content Rating             Genres  \
10472  1,000+     Free    0  Everyone            NaN  February 11, 2018   

      Last Updated Current Ver Android Ver  
10472       1.0.19  4.0 and up         NaN

Ở dòng 10472, giá trị cột `Category` bị thiếu, nên ta dịch các giá trị bắt đầu từ cột `Category` sang bên phải rồi điền giá trị cho cột `Category` là `NaN`

In [17]:
cols_list = app_df.columns.tolist()
i = len(cols_list) - 1
while i > 1:
    app_df.loc[bad_index, cols_list[i]] = app_df.loc[bad_index, cols_list[i - 1]]
    i -= 1
app_df.loc[bad_index, 'Category'] = np.nan

app_df.loc[bad_index]

App Category Rating Reviews  Size  \
10472  Life Made WI-Fi Touchscreen Photo Frame      NaN    1.9    19.0  3.0M   

      Installs  Type Price Content Rating Genres       Last Updated  \
10472   1,000+  Free     0       Everyone    NaN  February 11, 2018   

      Current Ver Android Ver  
10472      1.0.19  4.0 and up

Dữ liệu đã được sửa xong, bây giờ ta loại bỏ dòng bị lặp sau khi chỉnh sửa dòng này.

In [18]:
app_df = app_df.drop_duplicates()
get_app_df_shape()

[+] num_rows = 9660
[+] num_cols = 13


#### Tiếp tục khám phá dữ liệu

Ta tiếp tục liệt kê các kiểu dữ liệu khác nhau ở các cột có dtype là `object`.

In [19]:
def open_object_dtype(s):
    return set(s.apply(type).unique())

col_dtypes = app_df.dtypes

for col_name in col_dtypes.index:
    if col_dtypes[col_name] == 'object':
        print(f'[+] Column {col_name} dtypes: ')
        print(open_object_dtype(app_df[col_name]))
        print(app_df[col_name].unique())
        print()

[+] Column App dtypes: 
{<class 'str'>}
['Photo Editor & Candy Camera & Grid & ScrapBook'
 'U Launcher Lite – FREE Live Cool Themes, Hide Apps'
 'Sketch - Draw & Paint' ... 'Parkinson Exercices FR'
 'The SCP Foundation DB fr nn5n'
 'iHoroscope - 2018 Daily Horoscope & Astrology']

[+] Column Category dtypes: 
{<class 'str'>, <class 'float'>}
['ART_AND_DESIGN' 'AUTO_AND_VEHICLES' 'BEAUTY' 'BOOKS_AND_REFERENCE'
 'BUSINESS' 'COMICS' 'COMMUNICATION' 'DATING' 'EDUCATION' 'ENTERTAINMENT'
 'EVENTS' 'FINANCE' 'FOOD_AND_DRINK' 'HEALTH_AND_FITNESS' 'HOUSE_AND_HOME'
 'LIBRARIES_AND_DEMO' 'LIFESTYLE' 'GAME' 'FAMILY' 'MEDICAL' 'SOCIAL'
 'SHOPPING' 'PHOTOGRAPHY' 'SPORTS' 'TRAVEL_AND_LOCAL' 'TOOLS'
 'PERSONALIZATION' 'PRODUCTIVITY' 'PARENTING' 'WEATHER' 'VIDEO_PLAYERS'
 'NEWS_AND_MAGAZINES' 'MAPS_AND_NAVIGATION' nan]

[+] Column Rating dtypes: 
{<class 'str'>, <class 'float'>}
[4.1 4.7 4.5 4.3 4.4 3.8 4.2 4.6 3.2 4.0 4.8 3.9 4.9 3.6 3.7 nan 3.3 3.4
 3.5 3.1 5.0 2.6 3.0 2.5 2.8 2.7 1.0 1.9 2.9 2.3 2.2

#### Đánh giá kiểu dữ liệu hiện tại của các cột

Các cột đã có kiểu dữ liệu hợp lý:
- Cột `App` là tên ứng dụng, để kiểu `str` là hợp lý.
- Cột `Category` thể hiện các danh mục mà ứng dụng thuộc về, để kiểu `str` là hợp lý, chỉ có giá trị `NaN` là `float`.
- Cột `Genres` thể hiện thể loại mà ứng dụng thuộc về, để kiểu `str` là hợp lý, chỉ có giá trị `NaN` là `float`.
- Cột `Last Updated` là thời điểm cập nhật gần nhất của ứng dụng trên Play Store, đang ở dạng `str`, cần chuyển về `datetime`.
- Cột `Current Ver` thể hiện phiên bản hiện tại ứng dụng trên Play Store, để kiểu `str` là hợp lý, chỉ có giá trị `NaN` là `float`.
- Cột `Android Ver` thể hiện phiên bản Android phù hợp để cài ứng dụng, để kiểu `str` là hợp lý, chỉ có giá trị `NaN` là `float`.

Các cột cần xử lý tiếp:
- Cột `Rating` đang có giá trị `'1.9'` cần chuyển về `float`.
- Cột `Reviews` là số lượt đánh giá cho ứng dụng, đang ở kiểu `str` và `float`, giá trị này luôn là số nguyên nên cần chuyển về `int`.
- Cột `Size` là kích thước của ứng dụng, đang có đơn vị là KB (kết thúc bằng `k`) hoặc MB (kết thúc bằng `M`), hoặc `'Varies with device'`. Ta sẽ chuyển các kích thước về `float` và thống nhất đơn vị là KB, riêng giá trị `'Varies with device'` được gán bằng `nan`, để cột `Size` được xét là một cột numerical.
- Cột `Installs` và `Content Rating` là cột categorical với các category có thứ tự. Ta sử dụng `pd.Categorical` để định nghĩa thứ tự cho các category này.
- Cột `Type` thể hiện ứng dụng trả phí hoặc miễn phí, nên là kiểu `str`, chỉ có giá trị `NaN` là `float`. Ta cũng cần xem xét điền vào giá trị hợp lí cho giá trị `NaN` dựa vào giá trị cột `Price`.
- Cột `Price` thể hiện giá của ứng dụng, đang là `str`, có dạng dấu `$` sau đó là giá tiền, ngoại trừ giá trị `0`. Cần bỏ dấu `$` và chuyển cột này về `float`.

#### Chuyển dtype của cột `Rating` sang float

In [20]:
app_df = app_df.astype({'Rating': float})
app_df['Rating']

0        4.1
2        4.7
3        4.5
4        4.3
5        4.4
        ... 
10836    4.5
10837    5.0
10838    NaN
10839    4.5
10840    4.5
Name: Rating, Length: 9660, dtype: float64

#### Chuyển dtype của cột `Reviews` sang int

In [21]:
app_df = app_df.astype({'Reviews': int})
app_df['Reviews']

0           159
2         87510
3        215644
4           967
5           167
          ...  
10836        38
10837         4
10838         3
10839       114
10840    398307
Name: Reviews, Length: 9660, dtype: int64

#### Tiền xử lý cột `Size` và chuyển dtype cuả cột sang float

In [22]:
def to_kb(string):
    if string == 'Varies with device':
        return np.nan
    num = float(string[:-1])
    coef = 1 #Nếu chuỗi kết thúc bằng kí tự 'k', chỉ cần lấy phần số trong chuỗi
    if string[-1] == 'M':
        coef *= 1024 #Nếu chuỗi kết thúc bằng kí tự 'M', lấy phần số trong chuỗi nhân 1024
    return float(num * coef)

app_df['Size'] = app_df['Size'].apply(to_kb)
app_df['Size']

0        19456.0
2         8908.8
3        25600.0
4         2867.2
5         5734.4
          ...   
10836    54272.0
10837     3686.4
10838     9728.0
10839        NaN
10840    19456.0
Name: Size, Length: 9660, dtype: float64

#### Sử dụng `pd.Categorical` cho cột `Installs`

Quan sát các giá trị khác nhau trong cột `Installs`:

In [23]:
np.array(sorted(app_df['Installs'].unique()))

array(['0', '0+', '1+', '1,000+', '1,000,000+', '1,000,000,000+', '10+',
       '10,000+', '10,000,000+', '100+', '100,000+', '100,000,000+', '5+',
       '5,000+', '5,000,000+', '50+', '50,000+', '50,000,000+', '500+',
       '500,000+', '500,000,000+'], dtype='<U14')

Ta thấy rằng cột `Installs` thuộc loại cột categorical có thứ tự, nên ta sẽ định nghĩa thứ tự các giá trị này và áp dụng cho cột `Installs`.

In [24]:
installs_cate_list = ['0', '0+', '1+', '5+', '10+', '50+',
                     '100+', '500+', '1,000+', '5,000+', '10,000+', '50,000+',
                     '100,000+', '500,000+', '1,000,000+', '5,000,000+', '10,000,000+', '50,000,000+',
                     '100,000,000+', '500,000,000+', '1,000,000,000+']
app_df['Installs'] = pd.Categorical(app_df['Installs'], categories=installs_cate_list, ordered=True)
app_df['Installs']

0            10,000+
2         5,000,000+
3        50,000,000+
4           100,000+
5            50,000+
            ...     
10836         5,000+
10837           100+
10838         1,000+
10839         1,000+
10840    10,000,000+
Name: Installs, Length: 9660, dtype: category
Categories (21, object): ['0' < '0+' < '1+' < '5+' ... '50,000,000+' < '100,000,000+' < '500,000,000+' < '1,000,000,000+']

#### Sử dụng `pd.Categorical` cho cột `Content Rating`

Quan sát các giá trị khác nhau trong cột `Content Rating`:

In [25]:
np.array(sorted(app_df['Content Rating'].unique()))

array(['Adults only 18+', 'Everyone', 'Everyone 10+', 'Mature 17+',
       'Teen', 'Unrated'], dtype='<U15')

Tương tự cột `Installs`, ta thấy rằng cột `Content Rating` thuộc loại cột categorical có thứ tự, nên ta sẽ định nghĩa thứ tự các giá trị này và áp dụng cho cột `Content Rating`.

In [26]:
cr_cate_list = ['Everyone', 'Everyone 10+', 'Teen', 'Mature 17+', 'Adults only 18+']
app_df['Content Rating'] = pd.Categorical(app_df['Content Rating'], categories=cr_cate_list, ordered=True)
app_df['Content Rating']

0          Everyone
2          Everyone
3              Teen
4          Everyone
5          Everyone
            ...    
10836      Everyone
10837      Everyone
10838      Everyone
10839    Mature 17+
10840      Everyone
Name: Content Rating, Length: 9660, dtype: category
Categories (5, object): ['Everyone' < 'Everyone 10+' < 'Teen' < 'Mature 17+' < 'Adults only 18+']

#### Điền giá trị `nan` trong cột `Type`

Xét các dòng có giá trị `nan` tại cột `Type`:

In [27]:
index_check_row = app_df[app_df['Type'].isna()].index
app_df.loc[index_check_row]

App Category  Rating  Reviews  Size Installs Type  \
9148  Command & Conquer: Rivals   FAMILY     NaN        0   NaN        0  NaN   

     Price Content Rating    Genres   Last Updated         Current Ver  \
9148     0   Everyone 10+  Strategy  June 28, 2018  Varies with device   

             Android Ver  
9148  Varies with device

Nhận xét từ mô tả dữ liệu, cột `Type` có giá trị là `'Free'` tương đương với cột `Price` có giá trị là `0`. Do đó ta sẽ điền vào giá trị `Free` vào vị trí chứa giá trị `nan` này.

In [28]:
app_df['Type'].fillna('Free', inplace=True)
app_df.loc[index_check_row]

App Category  Rating  Reviews  Size Installs  \
9148  Command & Conquer: Rivals   FAMILY     NaN        0   NaN        0   

      Type Price Content Rating    Genres   Last Updated         Current Ver  \
9148  Free     0   Everyone 10+  Strategy  June 28, 2018  Varies with device   

             Android Ver  
9148  Varies with device

#### Tiền xử lý cột `Price` và chuyển dtype sang float

Cột này đang có dtype là `str`, mỗi giá trị có 2 dạng:

1. Là `0`
2. Là dạng dấu `$`, phía sau đó là giá tiền

Do đó, loại bỏ dấu `$`, sau đó chuyển sang `float`.

In [29]:
app_df['Price'] = app_df['Price'].replace(r'\$', '', regex=True)
app_df = app_df.astype({'Price': float})
app_df['Price'].sample(5)

9661    0.0
8435    0.0
5918    0.0
2536    0.0
9176    0.0
Name: Price, dtype: float64

#### Làm rõ ý nghĩa cột `Type` và cột `Price`

Theo mô tả dữ liệu, cột `Type` và cột `Price` có ý nghĩa như sau:
* **Type**: Kiểu ứng dụng miễn phí (Free) hay trả phí (Paid)
* **Price**: Số tiền trả phí của ứng dụng (tại thời điểm thu thập dữ liệu)

Như vậy, ta cần kiểm tra cột `Type` có giá trị là `'Free'` tương đương với cột `Price` có giá trị là 0, cột `Type` có giá trị là `'Paid'` tương đương với cột `Price` có giá trị lớn hơn 0.

Kiểm tra các giá trị trong cột `Price` khi cột `Type` có giá trị là `'Free'`:

In [30]:
app_df[app_df['Type'] == 'Free']['Price'].unique()

array([0.])

Kiểm tra có bao nhiêu giá trị bé hơn hoặc bằng 0 trong cột `Price` khi cột `Type` có giá trị là `'Paid'`:

In [31]:
price_unique_array = app_df.loc[app_df['Type'] == 'Paid', 'Price']
(price_unique_array <= 0).sum()

0

Vậy dữ liệu trong cột `Type` và cột `Price` đã đúng theo mô tả dữ liệu.

#### Chuyển dtype của cột `Last Updated` sang datetime

In [32]:
app_df['Last Updated'] = pd.to_datetime(app_df['Last Updated'], format="%B %d, %Y")
app_df['Last Updated']

0       2018-01-07
2       2018-08-01
3       2018-06-08
4       2018-06-20
5       2017-03-26
           ...    
10836   2017-07-25
10837   2018-07-06
10838   2017-01-20
10839   2015-01-19
10840   2018-07-25
Name: Last Updated, Length: 9660, dtype: datetime64[ns]

#### Kiểu dữ liệu hiện tại của các cột dữ liệu

In [33]:
app_df.dtypes

App                       object
Category                  object
Rating                   float64
Reviews                    int64
Size                     float64
Installs                category
Type                      object
Price                    float64
Content Rating          category
Genres                    object
Last Updated      datetime64[ns]
Current Ver               object
Android Ver               object
dtype: object

### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?

Ta có 5 cột `Rating`, `Reviews`, `Size`, `Price`, `Last Updated` thuộc nhóm các cột numerical (cột `Last Updated` đang có dtype là datetime nên xếp vào nhóm cột numerical luôn.). Với mỗi cột ta sẽ tính tỉ lệ phần trăm giá trị thiếu thuộc [0, 100], min, max.

In [34]:
def missing_ratio(s):
    return s.isna().mean() * 100

app_nume_cols = ['Rating', 'Reviews', 'Size', 'Price', 'Last Updated']
app_nume_col_profiles_df = app_df[app_nume_cols].agg([missing_ratio, pd.Series.min, pd.Series.max])
app_nume_col_profiles_df

Rating     Reviews           Size  Price  \
missing_ratio  15.144928         0.0      12.701863    0.0   
min             1.000000         0.0       8.500000    0.0   
max             5.000000  78158306.0  102400.000000  400.0   

                      Last Updated  
missing_ratio                  0.0  
min            2010-05-21 00:00:00  
max            2018-08-08 00:00:00

Các giá trị này không có gì bất thường.

### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?

Ta có 8 cột `App`, `Category`, `Installs`, `Type`, `Content Rating`, `Genres`, `Current Ver`, `Android Ver` thuộc nhóm các cột categorical. Với mỗi cột categorical ta sẽ tính tỉ lệ phần trăm giá trị thiếu thuộc [0, 100], số lượng giá trị khác nhau (không xét giá trị thiếu), danh sách các giá trị khác nhau (không xét giá trị thiếu).

In [35]:
def num_diff_vals(s):
    return s.nunique()

def diff_vals(s):
    return s.dropna().unique()

app_cate_cols = ['App', 'Category', 'Installs', 'Type', 'Content Rating', 'Genres', 'Current Ver', 'Android Ver']
app_cate_col_profiles_df = app_df[app_cate_cols].agg([missing_ratio, num_diff_vals, diff_vals])
app_cate_col_profiles_df 

App  \
missing_ratio                                                0.0   
num_diff_vals                                               9660   
diff_vals      [Photo Editor & Candy Camera & Grid & ScrapBoo...   

                                                        Category  \
missing_ratio                                           0.010352   
num_diff_vals                                                 33   
diff_vals      [ART_AND_DESIGN, AUTO_AND_VEHICLES, BEAUTY, BO...   

                                                        Installs  \
missing_ratio                                                0.0   
num_diff_vals                                                 21   
diff_vals      ['10,000+', '5,000,000+', '50,000,000+', '100,...   

                       Type  \
missing_ratio           0.0   
num_diff_vals             2   
diff_vals      [Free, Paid]   

                                                  Content Rating  \
missing_ratio                                           0.020704   
num_diff_vals                                                  5   
diff_vals      ['Everyone', 'Teen', 'Everyone 10+', 'Mature 1...   

                                                          Genres  \
missing_ratio                                           0.010352   
num_diff_vals                                                119   
diff_vals      [Art & Design, Art & Design;Creativity, Auto &...   

                                                     Current Ver  \
missing_ratio                                           0.082816   
num_diff_vals                                               2818   
diff_vals      [1.0.0, 1.2.4, Varies with device, 1.1, 1.0, 6...   

                                                     Android Ver  
missing_ratio                                           0.020704  
num_diff_vals                                                 33  
diff_vals      [4.0.3 and up, 4.2 and up, 4.4 and up, 2.3 and...

Các giá trị này không có gì bất thường.

## File `googleplaystore_user_reviews.csv`

### Đọc dữ liệu từ file csv vào dataframe

In [41]:
review_df = pd.read_csv('Data/googleplaystore_user_reviews.csv')
review_df.head()

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN  
3  Positive                0.40                0.875000  
4  Positive                1.00                0.300000

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [38]:
def get_review_df_shape():
    review_num_rows, review_num_cols = review_df.shape
    print(f'[+] num_rows = {review_num_rows}')
    print(f'[+] num_cols = {review_num_cols}')
    
get_review_df_shape()

[+] num_rows = 64295
[+] num_cols = 5


Dữ liệu có 64295 dòng và 5 cột.

### Mỗi cột có ý nghĩa gì?

* **App**: Tên ứng dụng
* **Translated_Review**: Đánh giá của ứng dụng (đã tiền xử lý và dịch sang Tiếng Anh)
* **Sentiment**: Mức độ cảm tình của đánh giá
* **Sentiment_Polarity**: Điểm Sentiment Polarity
* **Sentiment_Subjectivity**: Điểm Sentiment Subjec tivity

### Dữ liệu có các dòng bị lặp không?

In [39]:
review_df.duplicated().sum() > 0

True

### Tiền xử lý: Xóa các dòng bị lặp

In [42]:
review_df = review_df.drop_duplicates()
review_df.head()

App                                  Translated_Review  \
0  10 Best Foods for You  I like eat delicious food. That's I'm cooking ...   
1  10 Best Foods for You    This help eating healthy exercise regular basis   
2  10 Best Foods for You                                                NaN   
3  10 Best Foods for You         Works great especially going grocery store   
4  10 Best Foods for You                                       Best idea us   

  Sentiment  Sentiment_Polarity  Sentiment_Subjectivity  
0  Positive                1.00                0.533333  
1  Positive                0.25                0.288462  
2       NaN                 NaN                     NaN  
3  Positive                0.40                0.875000  
4  Positive                1.00                0.300000

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [43]:
col_dtypes = review_df.dtypes
col_dtypes

App                        object
Translated_Review          object
Sentiment                  object
Sentiment_Polarity        float64
Sentiment_Subjectivity    float64
dtype: object

Cột `Sentimemnt_Polarity` và `Sentiment_Subjectivity` là số điểm nên có kiểu `float64` là đúng. 3 cột còn lại đang có kiểu `object`, ta cần xem bên trong chúng có những kiểu gì.

In [46]:
for col_name in col_dtypes.index:
    if col_dtypes[col_name] == 'object':
        print(f'[+] Column {col_name} dtypes: ')
        print(open_object_dtype(review_df[col_name]))
        print(review_df[col_name].unique())
        print()

[+] Column App dtypes: 
{<class 'str'>}
['10 Best Foods for You' '104 找工作 - 找工作 找打工 找兼職 履歷健檢 履歷診療室' '11st' ...
 'Hotwire Hotel & Car Rental App' 'Housing-Real Estate & Property'
 'Houzz Interior Design Ideas']

[+] Column Translated_Review dtypes: 
{<class 'str'>, <class 'float'>}
['I like eat delicious food. That\'s I\'m cooking food myself, case "10 Best Foods" helps lot, also "Best Before (Shelf Life)"'
 'This help eating healthy exercise regular basis' nan ...
 'Dumb app, I wanted post property rent give option. Website work. Waste time space phone.'
 'I property business got link SMS happy performance still guys need raise bar guys Cheers'
 'Useless app, I searched flats kondapur, Hyderabad . None number reachable I know flats unavailable would keep posts active']

[+] Column Sentiment dtypes: 
{<class 'str'>, <class 'float'>}
['Positive' nan 'Neutral' 'Negative']



3 cột này đều có kiểu `str` và `float`, `float` là do có giá trị thiếu.

### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?

2 cột `Sentiment_Polarity` và `Sentiment_Subjectivity` thể hiện điểm số nên xếp vào nhóm numerical.

In [48]:
def missing_ratio(s):
    return s.isna().mean() * 100

review_nume_cols = ['Sentiment_Polarity', 'Sentiment_Subjectivity']
review_nume_col_profiles_df = review_df[review_nume_cols].agg([missing_ratio, pd.Series.min, pd.Series.max])
review_nume_col_profiles_df

Sentiment_Polarity  Sentiment_Subjectivity
missing_ratio            3.200887                3.200887
min                     -1.000000                0.000000
max                      1.000000                1.000000

Các giá trị không có gì bất thường.

### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?

3 cột còn lại là `App`, `Translated_Review` và `Sentiment` xếp vào loại categorical.

In [50]:
def num_diff_vals(s):
    return s.nunique()

def diff_vals(s):
    return s.dropna().unique()

review_cate_cols = ['App', 'Translated_Review', 'Sentiment']
review_cate_col_profiles_df = review_df[review_cate_cols].agg([missing_ratio, num_diff_vals, diff_vals])
review_cate_col_profiles_df 

App  \
missing_ratio                                                0.0   
num_diff_vals                                               1074   
diff_vals      [10 Best Foods for You, 104 找工作 - 找工作 找打工 找兼職 ...   

                                               Translated_Review  \
missing_ratio                                           3.217184   
num_diff_vals                                              27994   
diff_vals      [I like eat delicious food. That's I'm cooking...   

                                   Sentiment  
missing_ratio                       3.200887  
num_diff_vals                              3  
diff_vals      [Positive, Neutral, Negative]

Các giá trị không có gì bất thường.

# Đưa ra các câu hỏi có ý nghĩa cần trả lời

* Câu hỏi là gì?
* Nếu trả lời được câu hỏi thì sẽ có lợi ích gì?

## Với mỗi nhóm độ tuổi mà ứng dụng nhắm đến, top 5 chuyên mục có nhiều lượt tải về nhất là gì?

Trả lời được câu hỏi này sẽ giúp chúng ta biết được 5 chuyên mục yêu thích của từng nhóm độ tuổi, do đó nếu như ta muốn xây dựng 1 ứng dụng nhắm đến một lứa tuổi nào đó thì nên phát triển ý tưởng dựa trên chuyên mục nào để có thể bắt kịp xu hướng của lứa tuổi đó và có tiềm năng nhận được nhiều lượt tải về.

Kết quả lưu vào dataframe `category_crating_df` gồm 4 cột là `Content Rating`, `Category`, `Total Install`, và `Sample App` với ý nghĩa lần lượt là nhóm độ tuổi, chuyên mục, tổng số lượt tải, tên ứng dụng ngẫu nhiên thuộc chuyên mục đó.

### Tiền xử lý
- Các ứng dụng không có chuyên mục (giá trị `NaN` ở cột `Category`) thì thay bằng chuyên mục `OTHER`
- Cột `Installs` thể hiện số lượt tải ứng dụng, nhưng đang ở dạng cột mốc, vì vậy cần chuyển nó về dạng `int`, ví dụ `1,000+` sẽ thay thành `1001`

In [59]:
def change_installs_str_to_int(s):
    if s == '0':
        return 0
    return int(s[:-1].replace(',', '')) + 1

df = app_df[['Category', 'Content Rating', 'Installs']].copy()

df['Category'] = df['Category'].fillna('OTHER')
df['Installs'] = df['Installs'].apply(change_installs_str_to_int).astype('int')

df = df.groupby(['Content Rating', 'Category'])['Installs'].sum().reset_index()
df = df.groupby('Content Rating').apply(lambda x : x.sort_values(by = 'Installs', ascending = False).head(5)).reset_index(drop=False)
df

ValueError: cannot insert Content Rating, already exists

# Nhìn lại quá trình làm đồ án

* Mỗi thành viên: Đã gặp những khó khăn gì? (Hay mọi
chuyện đều thuận lợi)
* Mỗi thành viên: Có học được gì hữu ích? (Hay không
học được gì)
* Nhóm: Nếu có thêm thời gian thì sẽ làm gì?

# Tài liệu tham khảo